### PR TITLE
TYP: assorted annotations

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also avoid passing unnecesary arg to libreduction, and use a lower-level sanitize function in sanitize_index

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -1397,7 +1410,9 @@ def wrapper(arr, indexer, out, fill_value=np.nan):
     return wrapper
 
 
-def _take_2d_multi_object(arr, indexer, out, fill_value, mask_info):
+def _take_2d_multi_object(
+    arr: np.ndarray, indexer: np.ndarray, out: np.ndarray, fill_value, mask_info
+):
```
Comment (MarcoGorelli) [inline pandas/core/algorithms.py:1415]:
```suggestion
): -> None
```
(and similarly for some others, while we're here?)

Comment (jbrockmendel) [reply pandas/core/algorithms.py:1415]: im ambivalent about the `-> None` pattern

Comment (jreback) [reply pandas/core/algorithms.py:1415]: why these are ways to annotate
and make it super obvious what is happening 

we should always type void

Comment (jbrockmendel) [reply pandas/core/algorithms.py]: Updated + green

File: pandas/core/common.py [hunk 2]
```diff
@@ -37,7 +37,7 @@
 )
 from pandas.core.dtypes.generic import ABCExtensionArray, ABCIndex, ABCSeries
 from pandas.core.dtypes.inference import iterable_not_string
-from pandas.core.dtypes.missing import isna, isnull, notnull  # noqa
+from pandas.core.dtypes.missing import isna, isnull, notnull  # noqa:F401
```
Comment (jreback) [inline pandas/core/common.py:40]: we should remove these imports entirely if they are not being used (this was a holdover from a long time ago i think)

**Review by jreback — APPROVED** [review]
looks fine. small comments can be in a followon. 

cc @simonjayhawkins for typing

File: pandas/core/common.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/common.py:40]: sure, will add that to the next assorted-cleanups branch ive got going

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -527,7 +527,11 @@ def f(c, v):
 
 
 def factorize_array(
-    values: np.ndarray, na_sentinel: int = -1, size_hint=None, na_value=None, mask=None
+    values: np.ndarray,
+    na_sentinel: int = -1,
+    size_hint: Optional[int] = None,
+    na_value=None,
```
Comment (jreback) [inline pandas/core/algorithms.py:533]: Optional[Scalar]?

Comment (jbrockmendel) [reply pandas/core/algorithms.py:533]: we dont actually know its a scalar

File: pandas/core/common.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/common.py]: ended up adding that here

File: pandas/core/algorithms.py [hunk 4]
```diff
@@ -263,7 +263,7 @@ def _reconstruct_data(
     return values
 
 
-def _ensure_arraylike(values):
+def _ensure_arraylike(values) -> ArrayLike:
```
Comment (simonjayhawkins) [inline pandas/core/algorithms.py:266]: is_array_like is True for Series and Index so the return type should probably be AnyArrayLike

**Pull request merged**